File: internal/secret/secret.go
// Package secret holds credential material that must never be printed.
//
// A Secret is created from a reference such as "env:NAME" or "file:/path".
// The value is resolved on every Get call, so a rotated file (for example a
// projected Kubernetes ServiceAccount token) keeps working without a restart.
// String, MarshalJSON, Format and LogValue always produce "[REDACTED]".
package secret

import (
	"bytes"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"strings"
)

// Redacted is the only text a Secret ever prints.
const Redacted = "[REDACTED]"

// Secret is a reference to credential material.
// The zero value is empty and Get returns ErrEmpty.
type Secret struct {
	kind string // "env", "file", "literal" or "" (empty)
	ref  string // env var name, file path or the literal itself
}

// ErrEmpty is returned by Get on the zero Secret.
var ErrEmpty = errors.New("secret: empty")

// Parse turns a config reference into a Secret.
// Accepted forms: "env:NAME" and "file:/path". Anything else is rejected so a
// plain credential can never be written into the config file by mistake.
func Parse(ref string) (Secret, error) {
	switch {
	case strings.HasPrefix(ref, "env:"):
		name := strings.TrimPrefix(ref, "env:")
		if name == "" || strings.ContainsAny(name, " \t=") {
			return Secret{}, fmt.Errorf("secret: invalid environment variable name in %q", ref)
		}
		return Secret{kind: "env", ref: name}, nil
	case strings.HasPrefix(ref, "file:"):
		path := strings.TrimPrefix(ref, "file:")
		if path == "" {
			return Secret{}, fmt.Errorf("secret: empty file path in %q", ref)
		}
		return Secret{kind: "file", ref: path}, nil
	case ref == "":
		return Secret{}, errors.New("secret: empty reference")
	default:
		return Secret{}, errors.New("secret: value must be a reference of the form env:NAME or file:/path, not an inline secret")
	}
}

// MustParse is Parse for tests and constants; it panics on error.
func MustParse(ref string) Secret {
	s, err := Parse(ref)
	if err != nil {
		panic(err)
	}
	return s
}

// Literal wraps an in-memory value. It exists for tests and for values that
// arrive from somewhere other than the config file (an exchanged token, for
// example). Config loading never produces a literal Secret.
func Literal(value string) Secret {
	return Secret{kind: "literal", ref: value}
}

// IsZero reports whether the Secret holds no reference.
func (s Secret) IsZero() bool { return s.kind == "" }

// Ref describes where the secret comes from without revealing it,
// for example "env:JIRA_TOKEN" or "file:/secrets/token".
// A literal Secret reports "literal".
func (s Secret) Ref() string {
	switch s.kind {
	case "env", "file":
		return s.kind + ":" + s.ref
	case "literal":
		return "literal"
	default:
		return ""
	}
}

// Get resolves the secret. File secrets are read on every call and trailing
// whitespace is trimmed, matching how tokens are usually mounted.
func (s Secret) Get() ([]byte, error) {
	switch s.kind {
	case "env":
		v, ok := os.LookupEnv(s.ref)
		if !ok {
			return nil, fmt.Errorf("secret: environment variable %s is not set", s.ref)
		}
		if v == "" {
			return nil, fmt.Errorf("secret: environment variable %s is empty", s.ref)
		}
		return []byte(v), nil
	case "file":
		b, err := os.ReadFile(s.ref)
		if err != nil {
			return nil, fmt.Errorf("secret: read %s: %w", s.ref, err)
		}
		b = bytes.TrimRight(b, " \t\r\n")
		if len(b) == 0 {
			return nil, fmt.Errorf("secret: file %s is empty", s.ref)
		}
		return b, nil
	case "literal":
		return []byte(s.ref), nil
	default:
		return nil, ErrEmpty
	}
}

// GetString is Get as a string.
func (s Secret) GetString() (string, error) {
	b, err := s.Get()
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// String implements fmt.Stringer and never reveals the value.
func (s Secret) String() string { return Redacted }

// GoString keeps %#v from printing the struct fields.
func (s Secret) GoString() string { return "secret.Secret{" + Redacted + "}" }

// Format keeps every fmt verb from printing the struct fields.
func (s Secret) Format(f fmt.State, _ rune) { _, _ = f.Write([]byte(Redacted)) }

// MarshalJSON keeps encoding/json from printing the value.
func (s Secret) MarshalJSON() ([]byte, error) { return []byte(`"` + Redacted + `"`), nil }

// MarshalText keeps text encoders (YAML, logs) from printing the value.
func (s Secret) MarshalText() ([]byte, error) { return []byte(Redacted), nil }

// LogValue keeps log/slog from printing the value.
func (s Secret) LogValue() slog.Value { return slog.StringValue(Redacted) }

File: hallpass-py/src/hallpass/core/secret.py
"""Credential material that must never be printed.

A Secret is created from a reference such as ``env:NAME`` or ``file:/path``.
The value is resolved on every get() call, so a rotated file (a projected
Kubernetes ServiceAccount token, for example) keeps working without a
restart. str(), repr(), format() and pickling never reveal the value.
"""

from __future__ import annotations

import os
from typing import NoReturn

from hallpass.core.errors import go_quote, path_error_text

__all__ = ["REDACTED", "EmptySecretError", "Secret", "SecretError", "env", "file", "literal", "must_parse", "parse"]

REDACTED = "[REDACTED]"


class SecretError(ValueError):
    pass


class EmptySecretError(SecretError):
    """get() on the empty Secret (Go's secret.ErrEmpty)."""


class Secret:
    """A reference to credential material. ``Secret()`` is empty."""

    __slots__ = ("_kind", "_ref")

    def __init__(self, _kind: str = "", _ref: str = "") -> None:
        object.__setattr__(self, "_kind", _kind)
        object.__setattr__(self, "_ref", _ref)

    def __setattr__(self, name: str, value: object) -> NoReturn:
        raise AttributeError("Secret is immutable")

    def is_zero(self) -> bool:
        return self._kind == ""

    def ref(self) -> str:
        """Where the secret comes from without revealing it: "env:JIRA_TOKEN",
        "file:/secrets/token", or "literal"."""
        if self._kind in ("env", "file"):
            return f"{self._kind}:{self._ref}"
        if self._kind == "literal":
            return "literal"
        return ""

    def get(self) -> bytes:
        """Resolve the secret. File secrets are read on every call and
        trailing whitespace is trimmed, matching how tokens are mounted."""
        if self._kind == "env":
            v = os.environ.get(self._ref)
            if v is None:
                raise SecretError(f"secret: environment variable {self._ref} is not set")
            if v == "":
                raise SecretError(f"secret: environment variable {self._ref} is empty")
            return v.encode("utf-8", "surrogateescape")
        if self._kind == "file":
            try:
                with open(self._ref, "rb") as f:
                    b = f.read()
            except (OSError, ValueError) as e:
                # Go's os.ReadFile error, wrapped: "open /p: no such file or
                # directory"; a directory opens and then fails to read.
                op = "read" if isinstance(e, IsADirectoryError) else "open"
                raise SecretError(f"secret: read {self._ref}: {path_error_text(op, self._ref, e)}") from e
            b = b.rstrip(b" \t\r\n")
            if not b:
                raise SecretError(f"secret: file {self._ref} is empty")
            return b
        if self._kind == "literal":
            return self._ref.encode("utf-8", "surrogateescape")
        raise EmptySecretError("secret: empty")

    def get_string(self) -> str:
        return self.get().decode("utf-8", "surrogateescape")

    def __str__(self) -> str:
        return REDACTED

    def __repr__(self) -> str:
        return "Secret(" + REDACTED + ")"

    def __format__(self, spec: str) -> str:
        return REDACTED

    def __reduce__(self) -> NoReturn:
        raise TypeError("a Secret cannot be pickled")

    def __eq__(self, other: object) -> bool:
        return isinstance(other, Secret) and self._kind == other._kind and self._ref == other._ref

    def __hash__(self) -> int:
        return hash((self._kind, self._ref))

    def __bool__(self) -> bool:
        return not self.is_zero()


def parse(ref: str) -> Secret:
    """Turn a config reference into a Secret.

    Accepted forms: ``env:NAME`` and ``file:/path``. Anything else is
    rejected so a plain credential can never be written into the config
    file by mistake.
    """
    if ref.startswith("env:"):
        name = ref[len("env:") :]
        if name == "" or any(c in name for c in " \t="):
            raise SecretError(f"secret: invalid environment variable name in {go_quote(ref)}")
        return Secret("env", name)
    if ref.startswith("file:"):
        path = ref[len("file:") :]
        if path == "":
            raise SecretError(f"secret: empty file path in {go_quote(ref)}")
        return Secret("file", path)
    if ref == "":
        raise SecretError("secret: empty reference")
    raise SecretError("secret: value must be a reference of the form env:NAME or file:/path, not an inline secret")


def must_parse(ref: str) -> Secret:
    return parse(ref)


def literal(value: str) -> Secret:
    """Wrap an in-memory value: an exchanged token, or a credential an
    application passes in code. Config files never produce one."""
    return Secret("literal", value)


def env(name: str) -> Secret:
    """``env:NAME`` as a Secret, for connections configured in code."""
    return parse("env:" + name)


def file(path: str) -> Secret:
    """``file:/path`` as a Secret, for connections configured in code."""
    return parse("file:" + path)
